storage/src/linear/io_uring.rs

```diff
@@ -0,0 +1,811 @@
+        // we never take from the iterator without submitting it to the kernel
+        // or adding it to the backlog, so we don't need to check the iterator
+        self.backlog.is_empty() && self.outstanding.is_empty()
+    }
+
+    fn hashtable_insert(&mut self, entry: QueueEntry<'batch>) {
+        self.outstanding
+            .insert_unique(entry.hashtable_hash(), entry, QueueEntry::hashtable_hash);
+    }
+
+    fn hashtable_remove(&mut self, offset: u64) -> Option<QueueEntry<'batch>> {
+        match self.outstanding.entry(
+            offset,
+            |needle| needle.original_offset == offset,
+            QueueEntry::hashtable_hash,
+        ) {
+            HashTableEntry::Occupied(entry) => {
+                let (entry, _) = entry.remove();
+                Some(entry)
+            }
+            HashTableEntry::Vacant(_) => None,
+        }
+    }
+
+    fn pop_next_entry(&mut self) -> Option<QueueEntry<'batch>> {
+        if let Some(entry) = self.backlog.pop_front() {
+            Some(entry)
+        } else {
+            self.entries.next()
+        }
+    }
+
+    fn add_entry_to_backlog(&mut self, entry: QueueEntry<'batch>) {
+        if entry.offset == entry.original_offset {
+            // new entry, queue at the back
+            self.backlog.push_back(entry);
+        } else {
+            // partially written entry, re-queue at the front for priority
+            self.backlog.push_front(entry);
+        }
+    }
+
+    fn sync_completion_queue(&mut self) -> std::io::Result<()> {
+        trace!("syncing io-uring completion queue");
+        // `submit_and_wait` blocks until at least `want` (1) completion events
+        // are available in the completion queue. If there are already `want`
+        // events in the cq, this returns immediately.
+        //
+        // The return value is simply the number of SQEs copied to the kernel
+        // since the last time we called `submit` or `submit_and_wait` and is
+        // not particularly useful here.
+        //
+        // The kernel will spurrously interrupt this call if this thread needs
+        // to be woken up for an unrelated reason; we ignore those interrupts.
+        // The kernel will also return EBUSY if its internal submission queue is
+        // full, in which case we break out to process completions. If the CQ
+        // is empty, we may spin here until the syscall blocks.
+        match self.submitter.submit_and_wait(1) {
+            Ok(_) => {}
+            Err(ref err) if ignore_poll_error(err) => {}
+            Err(err) => return Err(err),
+        }
+
+        self.cq.sync();
+
+        Ok(())
+    }
+
+    fn enqueue_single_op(&mut self) -> EnqueueResult {
+        let Some(entry) = self.pop_next_entry() else {
+            return EnqueueResult::Empty;
+        };
+
+        // debug_assert_eq so it `Debug` prints the entry on failure
+        debug_assert_eq!(
+            self.hashtable_remove(entry.original_offset),
```

Comment:
I think this was supposed to always call `hashtable_remove` but I think the optimizer will remove this if debug_assertions is off, leaving those entries in the hash table.